src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokenCertificatesApiController.java

```diff
@@ -0,0 +1,95 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.openapi;
+
+import ee.ria.xroad.common.conf.serverconf.model.CertificateType;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.niis.xroad.restapi.converter.CertificateDetailsConverter;
+import org.niis.xroad.restapi.exceptions.ErrorDeviation;
+import org.niis.xroad.restapi.openapi.model.CertificateDetails;
+import org.niis.xroad.restapi.service.CertificateAlreadyExistsException;
+import org.niis.xroad.restapi.service.ClientNotFoundException;
+import org.niis.xroad.restapi.service.GlobalConfService;
+import org.niis.xroad.restapi.service.KeyNotFoundException;
+import org.niis.xroad.restapi.service.TokenCertificateService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.Resource;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * certificates api
+ */
+@Controller
+@RequestMapping("/api")
+@Slf4j
+@PreAuthorize("denyAll")
+public class TokenCertificatesApiController implements TokenCertificatesApi {
+    public static final String ERROR_INVALID_CERT_UPLOAD = "invalid_cert_upload";
+    public static final String ERROR_INVALID_CERT = "invalid_cert";
+
+    private final TokenCertificateService tokenCertificateService;
+    private final CertificateDetailsConverter certificateDetailsConverter;
+
+    @Autowired
+    public TokenCertificatesApiController(TokenCertificateService tokenCertificateService,
+            CertificateDetailsConverter certificateDetailsConverter) {
+        this.tokenCertificateService = tokenCertificateService;
+        this.certificateDetailsConverter = certificateDetailsConverter;
+    }
+
+    @Override
+    @PreAuthorize("hasAnyAuthority('IMPORT_AUTH_CERT', 'IMPORT_SIGN_CERT')")
+    public ResponseEntity<CertificateDetails> addCertificate(Resource certificateResource) {
+        byte[] certificateBytes;
+        try (InputStream is = certificateResource.getInputStream()) {
+            certificateBytes = IOUtils.toByteArray(is);
+        } catch (IOException ex) {
+            throw new BadRequestException("cannot read certificate data", ex,
+                    new ErrorDeviation(ERROR_INVALID_CERT_UPLOAD));
+        }
```

Comment:
I'm not sure if it fits in a converter since they are mainly used for OpenAPI/Core DTO -conversions. In this case we are just reading bytes from an InputStream so maybe a utility method would be suitable? Of course the existing `FormatUtils` does not sound right for this purpose.

Reply:
Created a new `ResourceUtils` utility class which now has a (awkwardly named) method `springResourceToBytesOrThrowBadRequest` in order to remove the duplicate code.

Reply:
There's also `ResourceUtils` class in common-util. I happened to add createAttachmentResourceResponse, which is close to this one, to ApiUtil: https://github.com/nordic-institute/X-Road/blob/XRDDEV-685-generate-csr/src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/ApiUtil.java#L92-L101

What do you think about moving this to ApiUtil instead?

In any case, since ResourceUtils/ApiUtil is for generic use, variable and especially error code naming should be generic instead of `invalid_cert_upload`

Reply:
This seems to be the last open review comment. I can handle the possible move to ApiUtil in XRDDEV-685 PR. But his part should be handled in the PR:

"ResourceUtils/ApiUtil is for generic use, variable and especially error code naming should be generic instead of invalid_cert_upload"

Reply:
Changed the error to a more generic one. I still left the ResourceUtil to be but you can go ahead and move the method into ApiUtils since it also fits there nicely 👍

Reply:
There's still `certificateBytes`.... :)

Reply:
Allllrighty then! Let's wait for the build to finish now 👷‍♂️